Faster complex models

KimMcAlpine opened the pull request:
Improve the performance of the pipeline's calculation of visibilities for complex models (models with multiple sources located away from the phase centre) as described in Jira ticket [SPR1-104](https://skaafrica.atlassian.net/browse/SPR1-104)

This is to allow the pipeline to incorporate these models when calculating the MeerKAT bandpass. The presence of sources away from the phase centre cause a 'ripple' modulation of the amplitude and phase of the visibilities which are more rapid on longer baselines. If not accounted for this could result in 'ripples' in the final computed bandpass. 

A few other minor changes are:
* If there is no default model for a source in the `/conf/sky_models` directory, the pipeline now uses the model provided by the target string
* The pipeline models are now provided for a MHz (not GHz) scale
* The spectrum of the dominant source in the model in now plotted in the `corrected_data` plots versus frequency in the cal report.

Commit:
Improve performance of complex models

This commit speeds up the performance of the complex, multi-source
sky model computation as suggested in jira ticket
https://skaafrica.atlassian.net/browse/SPR1-104.

This is accomplished by:
* computing the K-jones term per antenna, instead of per baseline
* calculating the product of uvw and lmn once and dividing by wavelength, instead of per channel
* using numba

Commit:
Convert flux density to float32 in complex models

This should slightly improve the performance of the model calculation

Commit:
Use flux model provided in target string

This commit addresses jira ticket https://skaafrica.atlassian.net/browse/SR-893.
Previously the pipeline used only its own internal calibrator models, stored in
/conf/sky_models. This PR allows the pipeline to search for an internal calibrator model,
if one isn't found it will use the model provided by the target string.

To enable this the format of the internal models is modified to better match
the format of  `katpoint.FluxDensityModel`, most notably it now stores model coefficients for the
MHz scale (previously it used the GHz scale). In the future the plan is that katpoint
will be flexible enough to handle both MHz and GHz scales allowing us to store/supply
the co-efficients in the form provided by the original source/reference.

When searching for a model file for a given target the code now searches for a file which matches
the name or the aliases of the target, this helps to slightly robustify the process used to find a model
file.

Commit:
Plot model spectrum in report

For plots of the spectra of gain calibrated targets include the spectrum
of the model of the source. This helps to validate that calibration has
corrected scaled the visibilites to the source model.

For sources with multiple sources in model only
plot the spectrum of the first source which is assumed to be the brightest and
dominant source in the field.

Commit:
Slightly streamline the uvw co-ordinate calculation

Commit:
Tests for complex models

bmerry reviewed (changes requested):
Here's a first batch of comments. I still need to review the report code and the tests.

bmerry reviewed (commented):
I've reviewed the report code too now. I think it probably makes sense to wait for my current comments to be addressed before reviewing the tests in case they cause the tests to change.

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+                for p in range(npols):
+                    for a in range(nants):
+                        phase = l * uvw[0, t, p, a] + m * uvw[1, t, p, a] + (n-1) * uvw[2, t, p, a]
+                        k_ant[t, c, p, a] = np.exp(2 * np.pi * 1j * phase / wl[c])
```
I've made the suggested changes

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+    Parameters
+    ----------
+    uvw : :class:`np.ndarray`, real, shape (3, ntimes, npols, nants)
+        uvw co-ordinates of antennas
```
Yes, I've removed the polarisation axis the UVW and K parameters

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+@numba.jit(nopython=True, parallel=True)
+def ant_to_bls(k_ant, ant1, ant2, k_bls):
+    """
+    Convert an array of per antenna complex gains to a per baseline array
```
I've incorporated this function into the `add_model_vis` function, so the docstring is slightly altered, but I have hyphenated "per antenna"

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+                for p in range(npols):
+                    for b in range(nbls):
+                        k_bls[t, c, p, b] = (k_ant[t, c, p, ant1[b]]
+                                             * np.conj(k_ant[t, c, p, ant2[b]]))
```
I've removed the polarisation axis from the model. I think its simplifies the code as we aren't really doing proper polarisation calculations at the moment. I've also updated the docstring to indicate that the model is calculated assuming that `S` is the  Stokes I flux density.

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+    Parameters
+    ----------
+    k_ant : :class:`np.ndarray`, complex, shape (ntimes, nchans, npols, nants)
+        array of per antenna quantities
```
This function as been subsumed into the `add_model_vis` function so the docstring no longer exists

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+    ant2 : :class:`np.ndarray`, int, shape (nbls)
+        index of second antenna of each baseline pair
+    k_bls : :class:`np.ndarray`, complex, shape (ntimes, nchans, npols, nbls)
+        array to update with per baseline gains
```
Docstring is now obsolete

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+    Parameters
+    ----------
+    K : :class:`np.ndarray`, complex, shape (ntimes, nchans, npols, nbls)
+        K-jones term per baseline
```
Capitalised

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+                for p in range(npols):
+                    for b in range(nbls):
+                        model[t, c, p, b] += K[t, c, p, b] * S[c]
+    return model
```
In my test notebook implementing in numba (with parallel =True) was roughly 20 x faster than doing it in numpy. 
I have implemented your suggestion of combining the `ant_to_bls` and `add_model_vis` steps into one function.

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+    :class: `np.ndarray`, complex, shape (ntimes, nchans, npols, nants)
+        K-Jones term per antenna
+    """
+    n = np.sqrt(1 - l**2 - m**2)
```
Done

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -849,84 +848,94 @@ def _create_model(self, max_offset=8., timestamps=None):
+                # calculate uvw, if it hasn't already been calculated
+                if self.uvw is None:
+                    uvw = self.target.uvw(self.antennas, self.timestamps, self.array_position)
+                    self.uvw = np.array(uvw, np.float32)
```
As discussed `self.target` is the `cbf_target` sensor which should be the phase centre of the array. The only possible wrinkle is there is an option to apply an offset to the pointing from the `cbf_target` but this is unlikely to be an option selected when observing a calibrator. So I'm leaving this as is for now.

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
+    """
+    Calculate K-Jones term per antenna
+
+    Calculate the K-Jones term, i.e the fourier transform phase term,
```
I just meant it to be a reminder that K-Jones means the phase of the 2D Fourier transform between visibility and sky brightness. Sometimes its hard to remember which of Jones matrix is which :). I've altered the docstring to hopefully be a bit clearer.

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -849,84 +848,94 @@ def _create_model(self, max_offset=8., timestamps=None):
+        min_freq = self.model_raw_params['min_freq']
+        max_freq = self.model_raw_params['max_freq']
+        if (min(min_freq) > min(self.channel_freqs / 1e6)) \
+                & (max(max_freq) < max(self.channel_freqs / 1e6)):
```
Yes you are correct, replaced with `or`

KimMcAlpine commented on katsdpcal/katsdpcal/pipelineprocs.py:
```diff
@@ -409,29 +411,44 @@ def get_model(name, lsm_dir_list=[]):
-                       ('RA', 'S24'), ('dRA', 'S8'), ('DEC', 'S24'), ('dDEC', 'S8'),
-                       ('a0', 'f8'), ('a1', 'f8'), ('a2', 'f8'), ('a3', 'f8'),
-                       ('fq', 'f8'), ('fu', 'f8'), ('fv', 'f8')]
         model_components = np.genfromtxt(model_file, delimiter=',', dtype=model_dtype)
```
I've implemented this, but with a small caveat. If you read in the file as a katpoint catalogue you can't store the catalogue directly in telstate,  this requires you to convert the catalogue to a list of  target description strings. However this process of converting to a description string causes a loss in precision in the coefficients in the flux model which only stores 4 significant figures i.e 34. 6776 becomes 34.68.  

To avoid this I'm reading in the file as a list of description strings, which gets stored in telstate, and then converting the model to a catalogue using the `add_model` function.

bmerry replied:
Maybe we should just fix katpoint so that it doesn't throw away precision? It's probably a one-line fix and it'll no doubt cause problems elsewhere too.

KimMcAlpine commented on katsdpcal/katsdpcal/pipelineprocs.py:
```diff
@@ -409,29 +411,44 @@ def get_model(name, lsm_dir_list=[]):
+            model_list += glob.glob('{0}/*{1}*.txt'.format(glob.os.path.abspath(lsm_dir), name))
         # ignore tilde ~ backup files
-        model_file_list = [f for f in model_file_list if f[-1] != '~']
+        model_list = [f for f in model_list if f[-1] != '~']
```
I've removed this

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -849,84 +848,94 @@ def _create_model(self, max_offset=8., timestamps=None):
+                    source_flux = katpoint.FluxDensityModel(0, 100.0e3, coefs=source_coefs)
+
+                    # katsdpcal flux model parameters referenced to GHz, not MHz
+                    # (so use frequencies in GHz)
```
removed comment

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -849,84 +848,94 @@ def _create_model(self, max_offset=8., timestamps=None):
+                    source_coefs = [source[a].item() for a in ['a0', 'a1', 'a2', 'a3', 'a4', 'a5']]
+                    source_flux = katpoint.FluxDensityModel(0, 100.0e3, coefs=source_coefs)
+                    # katsdpcal flux model parameters referenced to GHz, not MHz
+                    # (so use frequencies in GHz)
```
removed comment

KimMcAlpine commented on katsdpcal/katsdpcal/conf/sky_models/J0408-6545.txt:
```diff
@@ -1,4 +1,39 @@
-S0, P, 04:08:20.3788, 0, -65:45:09.08, 0, 1.3499, -1.0387, -0.3467, 0.0861, 0, 0, 0
+# The model below is bootstrapped with 1934-638, which is in turn bootstrapped with the Reynolds model in CASA 4.7.
+# See memo COMM-192 for full details
+# Polarization angle for PKS0407-65 is assumed -25 degrees. Fractional polarization of
```
I've included the polarisation in the model now

KimMcAlpine commented on katsdpcal/katsdpcal/pipelineprocs.py:
```diff
@@ -385,15 +385,17 @@ def get_baseline_mask(bls_lookup, ants, limits):
     ----------
-    name : str
-        name of source
+    name : :class:`katpoint.Target`
```
Changed

KimMcAlpine commented on katsdpcal/katsdpcal/conf/sky_models/3C286.txt:
```diff
@@ -1,6 +1,10 @@
+#    a0 = b0 - 3*b1 + 9*b2 - 27*b3; a1 = b1 - 6*b2 + 27*b3; a2 = b2 - 9*b3; a3 = b3
+#    ==> a0 = 0.1823, a1 = 1.4757, a2 = -0.4739, a3 = 0.0336
+
+S0, P, 13:31:08.3, 30:30:33, 400, 50e3, 0.1823, 1.4757, -0.4739, 0.0336, 0, 0, 1, 0, 0, 0
```
I've included the polarisation in the model now

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -849,84 +848,94 @@ def _create_model(self, max_offset=8., timestamps=None):
+                    ('a2' not in self.model_raw_params.dtype.names
+                     or self.model_raw_params['a2'] == 0) and
+                    ('a3' not in self.model_raw_params.dtype.names
+                     or self.model_raw_params['a3'] == 0)):
```
I'm using katpoint now, and checking that all the `coefs[1:6] == 0`

KimMcAlpine commented on katsdpcal/katsdpcal/report.py:
```diff
@@ -101,6 +103,36 @@ def utc_tstr(timestamp, day=False):
+    Returns
+    -------
+    flux_density : :class:`np.ndarray`
+        flux density in Jy
```
Done

Commit:
Convert cal models to katpoint description format

Commit:
Changes suggested by review

Commit:
Configure NUMBA for parallel use

Set the number of threads to num_workers
Set the threading layer to be fork and thread safe

Commit:
Merge branch 'master' into faster_complex_models

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,6 +578,124 @@ def normalise_complex(x):
     return norm_factor
 
 
+@numba.jit(nopython=True, parallel=True)
```
I'm setting the number of workers to the `self.num_workers` attribute of the Pipeline class. I have also set the numba `THREADING_LAYER` to `safe`. This should use the TBB  threading layer. I did a very simple test and I think TBB should work in the default container without any extra packages.

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -849,84 +848,94 @@ def _create_model(self, max_offset=8., timestamps=None):
+            self.logger.info(
+                '     A point source in the model has a model which '
+                'is not applicable for part of the band, setting model flux to 1')
+            self.model = np.array([1], np.float32)
```
I  think it could be quite dangerous to throw out  only some sources in the model. The overall model could be very strongly biased by one or two strong sources in the field and throwing these out could result in a very poor model, certainly it could be just as different to the true sky model  as just using the dominant source in the field. So I think that unless all the sources in the sky model are correct its probably a good default to  use only the dominant source.

I've modified it now so that it will default to the dominant source, but if this also has a model which is not valid in the given frequency range it will fall back to 1Jy

bmerry commented on katsdpcal/katsdpcal/test/test_control.py:
```diff
@@ -554,6 +558,7 @@ def make_vis(self, K, G, target, noise=np.array([])):
         vis = flux_density * np.exp(2j * np.pi * (K[pol1, ant1] - K[pol2, ant2]) * freqs) \
             * (G[pol1, ant1] * G[pol2, ant2].conj())
 
+        print(np.average(flux_density))
```
This looks like debug code you left in?

KimMcAlpine replied:
Oops, yes - I've removed it

bmerry commented on katsdpcal/katsdpcal/test/test_pipelineprocs.py:
```diff
@@ -177,3 +180,69 @@ def test_unexpected_parameters(self):
+    def setUp(self):
+        self.lsm_dir = tempfile.mkdtemp()
+        self.file_model = ('S0, radec, 08:25:26.87, -50:10:38.49, 800 1712'
+                           ' 5 6 7 8')
```
It's okay to make this a single line (the flake8 config sets a max length of 100), and I think more readable as one line. Also, shouldn't there be parentheses around the flux model?

KimMcAlpine replied:
I've made this a single line and included parentheses, but I think the parentheses are optional.

bmerry commented on katsdpcal/katsdpcal/test/test_pipelineprocs.py:
```diff
@@ -177,3 +180,69 @@ def test_unexpected_parameters(self):
+            fname = os.path.join(self.lsm_dir, name + '.txt')
+            f = open(fname, 'w')
+            f.write(self.file_model)
+            f.close()
```
It's preferred to use a `with` statement to ensure the file is closed; and since `self.file_model` doesn't itself have a newline, I would suggest using `print` rather than `write`:
```python
with open(fname, 'w') as f:
    print(self.file_model, file=f)
```

KimMcAlpine replied:
Done

bmerry commented on katsdpcal/katsdpcal/test/test_pipelineprocs.py:
```diff
@@ -177,3 +180,69 @@ def test_unexpected_parameters(self):
+            fname = os.path.join(self.lsm_dir, name + '.txt')
+            f = open(fname, 'w')
+            f.write(self.file_model)
+            f.close()
```
See comments above about writing model files.

KimMcAlpine replied:
Done

bmerry commented on katsdpcal/katsdpcal/test/test_pipelineprocs.py:
```diff
@@ -177,3 +180,69 @@ def test_unexpected_parameters(self):
+
+    def test_two_files(self):
+        """If two files match target name, confirm it returns model from one of them"""
+
```
Remove this blank line (PEP 257).

KimMcAlpine replied:
Removed

bmerry commented on katsdpcal/katsdpcal/test/test_calprocs.py:
```diff
@@ -415,6 +415,51 @@ def test_weighted(self):
+        expected_kant[:, 0] = np.exp(2 * np.pi * 1j * (3 + (np.sqrt(0.95) - 1) / 0.1))
+        expected_kant[:, 1] = np.exp(2 * np.pi * 1j * (0.75 + (np.sqrt(0.95) - 1) / 0.4))
+        expected_kant[1, 0, 3] = np.exp(2 * np.pi * 1j * (80 + (np.sqrt(0.95) - 1) / 0.1))
+        expected_kant[1, 1, 3] = np.exp(2 * np.pi * 1j * (20 + (np.sqrt(0.95) - 1) / 0.4))
```
`2 * np.pi * 1j` can be more compactly written as `2j * np.pi`.

KimMcAlpine replied:
Yes, I've changed this as suggested

bmerry commented on katsdpcal/katsdpcal/test/test_calprocs.py:
```diff
@@ -415,6 +415,51 @@ def test_weighted(self):
+        kant = calprocs.K_ant(uvw, l, m, wl, kant)
+
+        expected_kant = np.zeros((ntimes, nchans, nants), np.complex64)
+        expected_kant[:, 0] = np.exp(2 * np.pi * 1j * (3 + (np.sqrt(0.95) - 1) / 0.1))
```
Would be clearer (and easier to change) if one wrote `n = np.sqrt(1 - l * l - m * m)` where `l` and `m` are set up and then used `n`.

KimMcAlpine replied:
Yes, I've changed this as suggested

bmerry reviewed (changes requested):
It's looking much better - nice to see flux calculations finally leveraging katpoint. It's just minor niggles left.

Not entirely related to this PR, but I've heard on the grapevine that commissioning is standardising on J-names for target. Should the 3C* files be renamed to use those names?

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+                                                      self.cross_ant.bls_lookup[:, 1],
+                                                      S.astype(np.float32), complexmodel)
+
+            self.model = complexmodel[:, :, np.newaxis, :]
```
Comment added

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+                    *source.radec(), projection_type='SIN', coord_system='radec')
+
+                k_ant = calprocs.K_ant(self.uvw, l, m, wl, k_ant)
+                complexmodel = calprocs.add_model_vis(k_ant[:, :,  :],
```
Removed `[:, :, :]`

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
-                # source position
-                source_position = katpoint.construct_radec_target(source['RA'].item(),
-                                                                  source['DEC'].item())
+                    self.channel_freqs / 1.0e6)
```
Yes, its now on one line

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
-                # (so use frequencies in GHz)
-                #   currently using the same flux model for both polarisations
+                I_coefs = source.flux_model.coefs[:6]
+                source_flux = katpoint.FluxDensityModel(0, 100.0e3, coefs=I_coefs)
```
Changed to `source.flux_density(self.channel_freqs  / 1.0e6)`

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+                        '     Model: single point source, spectral model, average flux over '
+                        '{0:03.3f}-{1:03.3f} GHz: {2:03.4f} Jy'.format(
+                            self.channel_freqs[0] / 1.e9, self.channel_freqs[-1] / 1.e9,
+                            np.mean(self.model)))
```
Using MHz in the log message

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
             self.logger.info(
-                '     Model: {0} point sources'.format(len(np.atleast_1d(self.model_raw_params)),))
+                '     Model: {0} point sources'.format(
+                    len(self.model_raw_params),))
```
Changed to %-style formatting

bmerry replied:
Did you forget to push it?

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+                            self.model[0],))
+                else:
+                    # CASE B - Point source at the phase centre, with spectral slope
+                    source_flux = katpoint.FluxDensityModel(0, 100.0e3, coefs=I_coefs)
```
Changed to use the FluxDensityModel of the source in the catalogue

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+                    self.model = np.array([10.**I_coefs[0]], np.float32)
+                    self.logger.info(
+                        '     Model: single point source, flat spectrum, flux: {0:03.4f} Jy'.format(
+                            self.model[0],))
```
Fair point, I've removed this case. Now there are only two a single point source at the phase centre with a spectral model or a full model.

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
-                    '     Model: single point source, flat spectrum, flux: {0:03.4f} Jy'.format(
-                        self.model[0],))
+                    '     The  model is not valid '
+                    'for part of the band, setting model flux to 1')
```
Changed to be a warning

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+            self.model_raw_params = katpoint.Catalogue(self.model_raw_params.targets[0])
+            self.logger.info(
+                '     A source in the sky model is not valid'
+                ', selecting only the first source '
```
Changed to be a warning

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
+
+            freqs_MHz = freqs / 1e6
+            not_valid = (max(min_valid) > min(freqs_MHz)) or (min(max_valid) < max(freqs_MHz))
+            return ~not_valid
```
Changed as suggested

KimMcAlpine commented on katsdpcal/katsdpcal/scan.py:
```diff
@@ -963,89 +962,98 @@ def _create_model(self, max_offset=8., timestamps=None):
                                                  antenna=self.array_position)
 
+        def valid_allfreq(model_cat, freqs):
+            """ Check all models in model_cat are valid for the given freqs (in Hz) """
```
Spaces removed

KimMcAlpine commented on katsdpcal/katsdpcal/report.py:
```diff
@@ -101,6 +103,36 @@ def utc_tstr(timestamp, day=False):
+    """
+    Flux density of the first source in the target model
+
+    If there is no model, returns None
```
Added an empty line

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -583,29 +600,32 @@ def K_ant(uvw, l, m, wl, k_ant):
-    for a source with the given position (l, m) at the given wavelengths
+    Calculate the K-Jones term for a point source with the
+    given position (l, m) at the given wavelengths.
+    The K-Jones term is the phase term in the approximate
```
I've clarified that it is the geometric delay

KimMcAlpine commented on katsdpcal/katsdpcal/calprocs.py:
```diff
@@ -542,37 +543,53 @@ def k_fit(data, weights, corrprod_lookup, chans, refant=0, cross=True, chan_samp
+    if weights is None:
+        weights = np.ones_like(x, dtype=np.float32)
+    # ensure all NaN'ed data has zero weight
+    valid_weights = np.where(~np.isfinite(x), 0, weights)
```
Changed as suggested

Commit:
Changes suggested by review

Commit:
Change all calibrators to have J_xxxx names

Calibrators in /conf/sky_models now all have Jnames in line with
the style used by katsdpconfig/usr/catalogues

bmerry reviewed (changes requested):
Looks fine - just the one place where you commented to say you'd changed the logging but the change doesn't seem to be there.

Commit:
Use %-style logging as suggested by review

KimMcAlpine commented:
Sorry, just missed it. I've fixed it now

On Tue, Oct 22, 2019 at 11:33 AM Bruce Merry <notifications@github.com>
wrote:

> *@bmerry* requested changes on this pull request.
>
> Looks fine - just the one place where you commented to say you'd changed
> the logging but the change doesn't seem to be there.
> ------------------------------
>
> In katsdpcal/katsdpcal/scan.py
> <https://github.com/ska-sa/katsdppipelines/pull/404#discussion_r337414168>
> :
>
> >          # CASE C - Complex model requiring calculation via uvw coordinates ####
>          # If not one of the simple cases above, make a proper full model
>          else:
>              self.logger.info(
> -                '     Model: {0} point sources'.format(len(np.atleast_1d(self.model_raw_params)),))
> +                '     Model: {0} point sources'.format(
> +                    len(self.model_raw_params),))
>
> Did you forget to push it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ska-sa/katsdppipelines/pull/404?email_source=notifications&email_token=ACMD2WJTV2XUIRLOGIAHXFTQP3CFRA5CNFSM4IPPPJX2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIXQG6I#pullrequestreview-305070969>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACMD2WNOV7E4YY4TSAC732TQP3CFRANCNFSM4IPPPJXQ>
> .
>
